### ISSUE TITLE
[ Breaking Change Request ] Remove Observatory from the Dart VM

### ISSUE THREAD

**Issue by bkonyi** [dart-lang/sdk#50233]
The Observatory developer tooling was developed during the early days of the Dart project to allow for developers to debug, dive into CPU and memory performance characteristics, and inspect program structure of Dart programs. This tool was developed and maintained solely by Dart VM engineers without much input from UX designers, making it powerful for Dart SDK developers but difficult to use by the average Dart developer.

With the development of Dart DevTools by a dedicated team, it no longer makes sense to continue work on Observatory with the associated compatibility and maintenance burdens, especially now that Dart DevTools is at near feature parity with Observatory.

**Impact:**

Pre-Dart 3.0, users wishing to continue using Observatory will need to pass an `--enable-observatory` flag for Observatory to be served by the VM. For Dart 3.0, this flag will be removed and remaining Observatory users will be required to move to the new Dart DevTools developer tooling.

Dart DevTools is at near feature parity with Observatory and should meet the needs of the vast majority of Dart developers. Any missing functionality in DevTools that's critical to an individual's workflow should be documented in a new issue on the [DevTools issue tracker](https://github.com/flutter/devtools/issues/new).

**Proposal:**

- **Pre-Dart 3.0**: place Observatory behind `--enable-observatory`, making Observatory disabled by default by still available.
- **Dart 3.0**: remove the `--enable-observatory` flag and remove Observatory completely from the SDK.

cc @a-siva @kenzieschmoll @mit-mit

**Comment by a-siva** [comment]
//cc @zanderso

**Comment by zanderso** [comment]
The `--enable-observatory` flag should be made available before it is required. This is so that we can make changes to pass the flag without losing access to the Observatory.

**Comment by bkonyi** [comment]
> The `--enable-observatory` flag should be made available before it is required. This is so that we can make changes to pass the flag without losing access to the Observatory.

SGTM.

**Comment by bkonyi** [comment]
[CL adding support for `--[no-]serve-observatory` is up for review](https://dart-review.googlesource.com/c/sdk/+/267720). Default behavior currently is to implicitly set `--serve-observatory`.

**Comment by itsjustkevin** [comment]
@vsmenon @Hixie @grouma @zanderso it seems like the comments here have been addressed by @bkonyi .  How do we feel about this breaking change request?

**Comment by zanderso** [comment]
I'm not sure the timeline is achievable, but otherwise the plan sgtm. I also added https://github.com/flutter/devtools/issues/4711 to the tracker.

**Comment by a-siva** [comment]
lgtm with the caveat that the performance of devtools is acceptable to the engineers who currently use observatory

**Comment by vsmenon** [comment]
lgtm

**Comment by itsjustkevin** [comment]
@bkonyi do you find you have enough time to implement this?

**Comment by bkonyi** [comment]
The initial work needed to allow for disabling Observatory by default is [already up for review](https://dart-review.googlesource.com/c/sdk/+/267720). Once that change lands and `flutter_tools` has the right plumbing for `--serve-observatory`, we can go ahead and change the default behavior to not serve Observatory.

**Comment by zanderso** [comment]
@bkonyi `--serve-observatory` needs to be a flag that the Engine can pass to `Dart_SetVMFlags`. Is that what your change did?

**Comment by bkonyi** [comment]
> @bkonyi `--serve-observatory` needs to be a flag that the Engine can pass to `Dart_SetVMFlags`. Is that what your change did?

@zanderso I think we'll just need to make a change [here in the engine](https://github.com/flutter/engine/blob/f25656bf23ec61934b444a65fa85f1983922a618/runtime/dart_service_isolate.cc#L175) to allow for this option to be set.

**Comment by rmacnak-google** [comment]
Many gaps in functionality between Observatory and DevTools are filed under [this tag](https://github.com/flutter/devtools/issues?q=is%3Aissue+is%3Aopen+label%3A%22missing+core+functionality%22)

**Comment by srawlins** [comment]
Sorry for my ignorance, will we also remove the `--observe=` flag?

**Comment by itsjustkevin** [comment]
@bkonyi and @rmacnak-google are we planning on keeping this open until all issues in [this tag](https://github.com/flutter/devtools/issues?q=is%3Aissue+is%3Aopen+label%3A%22missing+core+functionality%22) are resolved?

**Comment by bkonyi** [comment]
> @bkonyi and @rmacnak-google are we planning on keeping this open until all issues in [this tag](https://github.com/flutter/devtools/issues?q=is%3Aissue+is%3Aopen+label%3A%22missing+core+functionality%22) are resolved?

I don't know if we'll end up porting everything under that tag as some of the functionality isn't critical / widely used, so we'll need to make a judgement call at some point.

> Sorry for my ignorance, will we also remove the --observe= flag?

@srawlins definitely not :-)

**Comment by bkonyi** [comment]
CL to disable serving Observatory by default is [up for review](https://dart-review.googlesource.com/c/sdk/+/287821).

**Comment by itsjustkevin** [comment]
Thanks @bkonyi.

CC: @XilaiZhang and @caseyhillers as this may impact flutter rolls.

**Comment by mit-mit** [comment]
@bkonyi looks like this landed, but got reverted?

**Comment by mit-mit** [comment]
OK, thanks

**Comment by itsjustkevin** [comment]
@bkonyi is this still something we are considering?

**Comment by maks** [comment]
While I was aware of the intent to direct Dart developers to use DevTools instead of Observatory I didn't realise there was also the intent to completely remove the Observatory 😞 

@mit-mit What is the recommendation for developers wanting to use functionality that was previously in Observatory that is not available in DevTools in the "VMTools" tab with "VM Developer Mode" enabled?

And given this has happened, why on earth is the above labelling used in DevTools?? I am pretty sure I am not doing any "VM development" yet I find it very necessary to use the tooling that is still available for Isolates in my Dart cli app project that involves using Isolates heavily.

**Comment by mit-mit** [comment]
@maks please file feature request issues for what functionality you are missing, describing the use case you would use them for.

**Comment by bkonyi** [comment]
> @mit-mit What is the recommendation for developers wanting to use functionality that was previously in Observatory that is not available in DevTools in the "VMTools" tab with "VM Developer Mode" enabled?
>
> And given this has happened, why on earth is the above labelling used in DevTools?? I am pretty sure I am not doing any "VM development" yet I find it very necessary to use the tooling that is still available for Isolates in my Dart cli app project that involves using Isolates heavily.

@maks We went with "VM Developer Mode" since the vast majority of the features behind that flag are only relevant to individuals working on the Dart VM. As Michael suggested, if there's any features that you use in Observatory that haven't been migrated to DevTools, please [file an issue in the DevTools repo](https://github.com/flutter/devtools/issues/new) describing the feature (screenshots from Observatory would be very helpful!) and how you use it in your workflow. Feel free to cc me on the issue.

Also, if you need access to Observatory for tooling that isn't available in DevTools yet, you can either pass `--serve-observatory` to `dart` or navigate to `http://<vm_service_uri>/_serveObservatory` to enable Observatory.

**Comment by itsjustkevin** [comment]
Checking in again with @Hixie @vsmenon and @grouma for this breaking change.

**Comment by Hixie** [comment]
I'm a bit confused as to what is being proposed here. Are `--enable-observatory` and `--serve-observatory` different? Does this actually remove observatory or will it still be there but less accessible?

**Comment by grouma** [comment]
ACX is completely agnostic to this change so SGTM.

**Comment by bkonyi** [comment]
> I'm a bit confused as to what is being proposed here. Are `--enable-observatory` and `--serve-observatory` different? Does this actually remove observatory or will it still be there but less accessible?

We went with `--serve-observatory`, and that change landed a few months ago. I think we're looking for approval for the eventual removal of Observatory from the Dart SDK.

**Comment by Hixie** [comment]
I'm definitely on board with the _eventual_ removal of the Observatory, absolutely. I think that's a goal to drive to. It's not clear to me that we're there yet though; from what I hear, there are still some features that developers need that are only available via the Observatory?

**Comment by bkonyi** [comment]
That's correct. We've still got some long tail features that need to be migrated (e.g, support for viewing profiler samples in the timeline viewer), but we're in a pretty good state with DevTools at this point. [We're tracking the remaining issues that we're aware of on the DevTools issue tracker](https://github.com/flutter/devtools/issues?q=is%3Aissue+is%3Aopen+label%3A%22observatory+migration%22).

We're also not planning on migrating 100% of Observatory to DevTools. Some Observatory functionality is no longer relevant to the VM team (e.g., the heap map) or is super niche (e.g., setting + displaying column breakpoints), so those features won't be migrated or will be added later.

> from what I hear, there are still some features that developers need that are only available via the Observatory?

If you're aware of any features that are important to developers on the Dash team or externally, I'd love to hear from you :) It's been difficult getting feedback on what Observatory features are still useful and important to the Dash team. Putting Observatory behind a flag was partly done to encourage people to use DevTools and file issues, but we've only seen a couple of feature requests filed since then.

**Comment by Hixie** [comment]
https://github.com/dart-lang/sdk/issues/50233#issuecomment-1689278462 (above) is the main feedback I've seen so far.

**Comment by bkonyi** [comment]
Ah okay, good to know. We haven't seen any issues filed related to https://github.com/dart-lang/sdk/issues/50233#issuecomment-1689278462, so it's hard to take action on anything at this point without more details.

**Comment by julemand101** [comment]
@bkonyi I think this issue should also be put on the list: https://github.com/flutter/devtools/issues/1941

Rather annoying that DevTools hides the hot reload if it does not detect Flutter. If nothing else, please at least have a way to force the buttons with a "I know what I am doing" message.

**Comment by bkonyi** [comment]
@julemand101 thanks for the heads up! I've added https://github.com/flutter/devtools/issues/1941 to the work that needs to be done before Observatory is removed.

**Comment by itsjustkevin** [comment]
@bkonyi checking in on this breaking change.

**Comment by bkonyi** [comment]
@itsjustkevin we still have functionality to migrate, so there's no plan to move forward with this for at least another quarter or two.

**Comment by itsjustkevin** [comment]
Thanks @bkonyi!  @vsmenon any opinions?  I would like to mark this breaking change as approved based on the previous comments and work already done.

**Comment by itsjustkevin** [comment]
@bkonyi checking in on this again, is this still WIP?

**Comment by bkonyi** [comment]
Observatory was removed from the shipped Dart SDK in 60b217ff2d7625ee2024f21c04188479fb87908f.
